Voline Rework

caelwarner opened the pull request:
a pr, just for you swan

caelwarner committed:
Reworked Voline

- Voline now eat magma cream and gold items
- They will grow in size as they eat until they can no longer move
- After eating gold items, they spit out molten gold clusters
- Added molten gold cluster
- 9 molten gold clusters can be crafted into 2 gold ingots

caelwarner committed:
Added EatItemsGoal

- Forgot to commit this file on the first commit

SwanX1 commented on src/main/java/com/nekomaster1000/infernalexp/client/entity/model/VolineModel.java:
```diff
@@ -76,6 +76,11 @@ public void setRotationAngles(T entityIn, float limbSwing, float limbSwingAmount
 		this.body.rotateAngleX = -MathHelper.abs(MathHelper.cos(limbSwing * 0.4662F) * 1.4F * limbSwingAmount);
 
+		if (entityIn.isEating()) {
+			System.out.println("EATING");
```
```suggestion
```
no

caelwarner replied:
hahaha oops

SwanX1 commented on src/main/java/com/nekomaster1000/infernalexp/client/entity/render/VolineGlowLayer.java:
```diff
@@ -1,25 +1,42 @@
     }
 
+    public RenderType getVolineRenderType(VolineEntity entity) {
+        return entity.getAttributeValue(Attributes.MOVEMENT_SPEED) <= 0 ? TIRED_RENDER_TYPE : RENDER_TYPE;
```
How can movement speed be less than 0
```suggestion
        return entity.getAttributeValue(Attributes.MOVEMENT_SPEED) == 0 ? TIRED_RENDER_TYPE : RENDER_TYPE;
```

caelwarner replied:
it's just a precautionary. no harm in making it <= just in case it happens to be below zero

SwanX1 commented on src/main/java/com/nekomaster1000/infernalexp/client/entity/render/VolineRenderer.java:
```diff
@@ -1,24 +1,38 @@
     @Override
     public ResourceLocation getEntityTexture(VolineEntity entity) {
-        return TEXTURE;
+        return entity.getAttributeValue(Attributes.MOVEMENT_SPEED) <= 0 ? TIRED_TEXTURE : TEXTURE;
```
How can movement speed be less than 0
```suggestion
        return entity.getAttributeValue(Attributes.MOVEMENT_SPEED) == 0 ? TIRED_TEXTURE : TEXTURE;
```

caelwarner replied:
it's just a precautionary. no harm in making it <= just in case it happens to be below zero

SwanX1 commented on src/main/java/com/nekomaster1000/infernalexp/entities/VolineEntity.java:
```diff
@@ -1,30 +1,84 @@
+//                    Items.GOLDEN_HOE, Items.GOLDEN_SWORD, Items.GOLDEN_SHOVEL, Items.GOLDEN_HELMET, Items.GOLDEN_CHESTPLATE, Items.GOLDEN_LEGGINGS, Items.GOLDEN_BOOTS)
+//    );
+
+    public static final Map<Item, Integer> EAT_ITEMS = new HashMap<Item, Integer>() {{
+        put(Items.MAGMA_CREAM, 0);
+        put(Items.GOLD_NUGGET, 0);
+        put(Items.GOLD_INGOT, 4);
+        put(Items.GOLD_BLOCK, 22);
+        put(Items.GOLD_ORE, 4);
+        put(Items.NETHER_GOLD_ORE, 4);
+        put(Items.GOLDEN_AXE, 9);
+        put(Items.GOLDEN_PICKAXE, 9);
+        put(Items.GOLDEN_SWORD, 5);
+        put(Items.GOLDEN_HOE, 5);
+        put(Items.GOLDEN_SHOVEL, 2);
+        put(Items.GOLDEN_HELMET, 11);
+        put(Items.GOLDEN_CHESTPLATE, 20);
+        put(Items.GOLDEN_LEGGINGS, 18);
+        put(Items.GOLDEN_BOOTS, 13);
+    }};
```
This should be converted into a tag or recipe type for modpacks to customize.

caelwarner replied:
Is there anyway for tags to hold some kind of value?

SwanX1 replied:
A loot table then, kinda like piglin bartering?
See: https://mcasset.cloud/1.16.5/data/minecraft/loot_tables/gameplay/piglin_bartering.json

SwanX1 commented on src/main/java/com/nekomaster1000/infernalexp/entities/VolineEntity.java:
```diff
@@ -39,21 +93,102 @@ public VolineEntity(EntityType<? extends MonsterEntity> type, World worldIn) {
+    }
+
+    public void setVolineSize(float size) {
+        if (size > 2.0F) {
+            size = 2.0F;
+        }
```
```suggestion
        size = Math.min(size, 2.0F);
```

SwanX1 commented on src/main/java/com/nekomaster1000/infernalexp/entities/VolineEntity.java:
```diff
@@ -39,21 +93,102 @@ public VolineEntity(EntityType<? extends MonsterEntity> type, World worldIn) {
+
+    @Override
+    public void readAdditional(CompoundNBT compound) {
+        float size = compound.getFloat("Size");
+        if (size < 1) {
+            size = 1;
+        }
```
```suggestion
        float size = Math.max(compound.getFloat("Size"), 1.0F);
```

SwanX1 commented on src/main/java/com/nekomaster1000/infernalexp/entities/VolineEntity.java:
```diff
@@ -80,5 +215,51 @@ public boolean isImmuneToFire() {
+        } else {
+            super.handleStatusUpdate(id);
+        }
+    }
+    public static class VolineEatItemsGoal extends EatItemsGoal<VolineEntity> {
```
Just a thematic suggestion, not needed
```suggestion
    }
    
    public static class VolineEatItemsGoal extends EatItemsGoal<VolineEntity> {
```

SwanX1 commented on src/main/java/com/nekomaster1000/infernalexp/entities/VolineEntity.java:
```diff
@@ -1,30 +1,84 @@
+    private static final DataParameter<Float> VOLINE_SIZE = EntityDataManager.createKey(VolineEntity.class, DataSerializers.FLOAT);
+    private boolean isEating;
+
+    private final Goal lookAtPlayer = new LookAtGoal(this, PlayerEntity.class, 8.0f);
```
```suggestion
    private final Goal lookAtPlayer = new LookAtGoal(this, PlayerEntity.class, 8.0F);
```

SwanX1 commented on src/main/java/com/nekomaster1000/infernalexp/entities/VolineEntity.java:
```diff
@@ -39,21 +93,102 @@ public VolineEntity(EntityType<? extends MonsterEntity> type, World worldIn) {
         //this.goalSelector.addGoal(0, new TemptGoal(this, 0.6D, TEMPTATION_ITEMS, false));
-        this.goalSelector.addGoal(0, new EatItemGoal<>(this, Items.MAGMA_CREAM, 32.0d, 1.0d));
-        this.goalSelector.addGoal(1, new MeleeAttackGoal(this, 0.6D, true));
+        this.goalSelector.addGoal(0, new VolineEatItemsGoal(this, EAT_ITEMS, 32.0d, getAttributeValue(Attributes.MOVEMENT_SPEED) * 2.0D));
```
```suggestion
        this.goalSelector.addGoal(0, new VolineEatItemsGoal(this, EAT_ITEMS, 32.0D, getAttributeValue(Attributes.MOVEMENT_SPEED) * 2.0D));
```

SwanX1 commented on src/main/java/com/nekomaster1000/infernalexp/entities/ai/EatItemsGoal.java:
```diff
@@ -0,0 +1,110 @@
+    @Override
+    public boolean shouldExecute() {
+        if (this.itemInstance == null) {
+            List<ItemEntity> list = this.entityIn.world.getEntitiesWithinAABB(ItemEntity.class, this.entityIn.getBoundingBox().grow(this.range, 3.0d, this.range));
```
```suggestion
            List<ItemEntity> list = this.entityIn.world.getEntitiesWithinAABB(ItemEntity.class, this.entityIn.getBoundingBox().grow(this.range, 3.0D, this.range));
```

SwanX1 commented on src/main/java/com/nekomaster1000/infernalexp/entities/ai/EatItemsGoal.java:
```diff
@@ -0,0 +1,110 @@
+    @Override
+    public void tick() {
+        if (eatDelay <= 0) {
+            if (this.entityIn.getDistanceSq(itemInstance) < 2.0d) {
```
```suggestion
            if (this.entityIn.getDistanceSq(itemInstance) < 2.0D) {
```

SwanX1 commented on src/main/java/com/nekomaster1000/infernalexp/entities/ai/EatItemsGoal.java:
```diff
@@ -0,0 +1,110 @@
+    public void tick() {
+        if (eatDelay <= 0) {
+            if (this.entityIn.getDistanceSq(itemInstance) < 2.0d) {
+                this.navigation.setSpeed(0.0d);
```
```suggestion
                this.navigation.setSpeed(0.0D);
```

SwanX1 commented on src/main/java/com/nekomaster1000/infernalexp/entities/ai/EatItemsGoal.java:
```diff
@@ -0,0 +1,110 @@
+
+                    } else if (eatTime % 3 == 0) {
+                        entityIn.lookAt(EntityAnchorArgument.Type.EYES, itemInstance.getPositionVec());
+                        entityIn.playSound(SoundEvents.ENTITY_GENERIC_EAT, 0.9f, 1.0f);
```
```suggestion
                        entityIn.playSound(SoundEvents.ENTITY_GENERIC_EAT, 0.9F, 1.0F);
```

caelwarner committed:
Small code changes

- Removed some debug code
- Made number suffixes uppercase
- Replaced some code with Math.max() and Math.min()

caelwarner committed:
Added JSON file to configure Voline eating mechanic

- Added voline_eat_table.json
- Learnt how minecraft interfaces with json files

SwanX1 commented on src/main/java/com/nekomaster1000/infernalexp/data/VolineEatTable.java:
```diff
@@ -0,0 +1,75 @@
+	protected void apply(Map<ResourceLocation, JsonElement> objectIn, IResourceManager resourceManagerIn, IProfiler profilerIn) {
+		ResourceLocation resourceLocation = new ResourceLocation(InfernalExpansion.MOD_ID, "custom/voline_eat_table.json");
+
+		try {
+			for (IResource iResource : resourceManagerIn.getAllResources(resourceLocation)) {
+				try (Reader reader = new BufferedReader(new InputStreamReader(iResource.getInputStream(), StandardCharsets.UTF_8))) {
+					JsonObject jsonObject = JSONUtils.fromJson(GSON_INSTANCE, reader, JsonObject.class);
+
+					if (jsonObject != null) {
+
+						for (JsonElement entry : jsonObject.getAsJsonArray("entries")) {
+
+							VOLINE_EAT_TABLE.put(ForgeRegistries.ITEMS.getValue(new ResourceLocation(entry.getAsJsonObject().get("accepted_item").getAsString())),
+									new HashMap<Item, Integer>() {{
+										for (JsonElement item : entry.getAsJsonObject().getAsJsonArray("returned_items")) {
+											put(ForgeRegistries.ITEMS.getValue(new ResourceLocation(item.getAsJsonObject().get("item").getAsString())),
+													item.getAsJsonObject().get("amount").getAsInt());
+										}
+							}});
+						}
+					}
+
+				} catch (RuntimeException | IOException exception) {
+					InfernalExpansion.LOGGER.error("Couldn't read voline eat table list {} in data pack {}", resourceLocation, iResource.getPackName(), exception);
+
+				} finally {
+					IOUtils.closeQuietly(iResource);
+				}
+			}
+		} catch (IOException exception) {
+			InfernalExpansion.LOGGER.error("Couldn't read voline eat table from {}", resourceLocation, exception);
+		}
+
+	}
+
+	public Map<Item, Map<Item, Integer>> getVolineEatTable() {
+		return VOLINE_EAT_TABLE;
+	}
+
+}
```
tabs bruh

SwanX1 commented on src/main/java/com/nekomaster1000/infernalexp/data/VolineEatTable.java:
```diff
@@ -0,0 +1,75 @@
+	protected void apply(Map<ResourceLocation, JsonElement> objectIn, IResourceManager resourceManagerIn, IProfiler profilerIn) {
+		ResourceLocation resourceLocation = new ResourceLocation(InfernalExpansion.MOD_ID, "custom/voline_eat_table.json");
+
+		try {
+			for (IResource iResource : resourceManagerIn.getAllResources(resourceLocation)) {
+				try (Reader reader = new BufferedReader(new InputStreamReader(iResource.getInputStream(), StandardCharsets.UTF_8))) {
+					JsonObject jsonObject = JSONUtils.fromJson(GSON_INSTANCE, reader, JsonObject.class);
+
+					if (jsonObject != null) {
+
+						for (JsonElement entry : jsonObject.getAsJsonArray("entries")) {
+
+							VOLINE_EAT_TABLE.put(ForgeRegistries.ITEMS.getValue(new ResourceLocation(entry.getAsJsonObject().get("accepted_item").getAsString())),
+									new HashMap<Item, Integer>() {{
+										for (JsonElement item : entry.getAsJsonObject().getAsJsonArray("returned_items")) {
+											put(ForgeRegistries.ITEMS.getValue(new ResourceLocation(item.getAsJsonObject().get("item").getAsString())),
+													item.getAsJsonObject().get("amount").getAsInt());
+										}
+							}});
+						}
+					}
+
+				} catch (RuntimeException | IOException exception) {
+					InfernalExpansion.LOGGER.error("Couldn't read voline eat table list {} in data pack {}", resourceLocation, iResource.getPackName(), exception);
+
+				} finally {
+					IOUtils.closeQuietly(iResource);
+				}
+			}
+		} catch (IOException exception) {
+			InfernalExpansion.LOGGER.error("Couldn't read voline eat table from {}", resourceLocation, exception);
+		}
```
Weird unnecessary line breaks

SwanX1 reviewed (approved):
All in all, looks good to me, just need to make an entry in the wiki documenting the new voline eat table, and we should be all set!
